translator/trace/protospan_to_jaegerthrift.go

```diff
@@ -0,0 +1,321 @@
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+)
+
+// OCProtoToJaegerThrift translates OpenCensus trace data into the Jaeger Thrift format.
+func OCProtoToJaegerThrift(ocBatch *agenttracepb.ExportTraceServiceRequest) (*jaeger.Batch, error) {
+	jb := &jaeger.Batch{
+		Process: ocNodeToJaegerProcess(ocBatch.Node),
+		Spans:   ocSpansToJaegerSpans(ocBatch.Spans),
+	}
+	return jb, nil
+}
+
+func ocNodeToJaegerProcess(node *commonpb.Node) *jaeger.Process {
+	if node == nil {
+		return nil
+	}
+
+	var jTags []*jaeger.Tag
+	nodeAttribsLen := len(node.Attributes)
+	if nodeAttribsLen > 0 {
+		jTags = make([]*jaeger.Tag, 0, nodeAttribsLen)
+		for k, v := range node.Attributes {
+			str := v
+			jTag := &jaeger.Tag{
+				Key:   k,
+				VType: jaeger.TagType_STRING,
+				VStr:  &str,
+			}
+			jTags = append(jTags, jTag)
+		}
+	}
+
+	if node.Identifier != nil && node.Identifier.HostName != "" {
+		hostTag := &jaeger.Tag{
+			Key:   "hostname",
+			VType: jaeger.TagType_STRING,
+			VStr:  &node.Identifier.HostName,
+		}
+		jTags = append(jTags, hostTag)
+	}
+
+	// Add OpenCensus library information as tags if available
+	ocLib := node.LibraryInfo
+	if ocLib != nil {
+		// Only add language if specified
+		if ocLib.Language != commonpb.LibraryInfo_LANGUAGE_UNSPECIFIED {
+			languageStr := ocLib.Language.String()
+			languageTag := &jaeger.Tag{
+				Key:   "opencensus.language",
+				VType: jaeger.TagType_STRING,
+				VStr:  &languageStr,
+			}
+			jTags = append(jTags, languageTag)
+		}
+		if ocLib.ExporterVersion != "" {
+			exporterTag := &jaeger.Tag{
+				Key:   "opencensus.exporterversion",
+				VType: jaeger.TagType_STRING,
+				VStr:  &ocLib.ExporterVersion,
+			}
+			jTags = append(jTags, exporterTag)
+		}
+		if ocLib.CoreLibraryVersion != "" {
+			exporterTag := &jaeger.Tag{
+				Key:   "opencensus.corelibversion",
+				VType: jaeger.TagType_STRING,
+				VStr:  &ocLib.CoreLibraryVersion,
+			}
+			jTags = append(jTags, exporterTag)
+		}
+	}
+
+	var serviceName string
+	if node.ServiceInfo != nil && node.ServiceInfo.Name != "" {
+		serviceName = node.ServiceInfo.Name
+	}
+
+	if serviceName == "" && len(jTags) == 0 {
+		// No info to put in the process...
+		return nil
+	}
+
+	jProc := &jaeger.Process{
+		ServiceName: serviceName,
+		Tags:        jTags,
+	}
+
+	return jProc
+}
+
+func ocSpansToJaegerSpans(ocSpans []*tracepb.Span) []*jaeger.Span {
+	if ocSpans == nil {
+		return nil
+	}
+
+	// Pre-allocate assuming that few, if any spans, are nil.
+	jSpans := make([]*jaeger.Span, 0, len(ocSpans))
+	for _, ocSpan := range ocSpans {
+		startTime := timestampToEpochMicroseconds(ocSpan.StartTime)
+		jSpan := &jaeger.Span{
+			SpanId:        ocIDBytesToJaegerID(ocSpan.SpanId),
+			ParentSpanId:  ocIDBytesToJaegerID(ocSpan.ParentSpanId),
+			OperationName: truncableStringToStr(ocSpan.Name),
+			References:    ocLinksToJaegerReferences(ocSpan.Links),
+			// Flags: TODO (@pjanotti) at first nothing matches to it.
```

Comment:
I don't understand this comment.

Reply:
Jaeger uses this as a bit flags but I didn't find the related/equivalent in OC proto, so it seems that the info that they carry is lost. I will reword it.